File: modules/CRM/Lead/Lead.py
class DocType:
  def __init__(self, doc, doclist):
    self.doc = doc
    self.doclist = doclist
  
  #def autoname(self):
    #self.doc.name = make_autoname(self.doc.lead_name+' / '+self.doc.company_name)

  # Gets states belonging to country selected
  # =====================================================================
  def check_state(self):
    return NEWLINE.join([i[0] for i in sql("select state_name from `tabState` where `tabState`.country='%s' " % self.doc.country)])
  
  # Get item detail (will be removed later)
  #==================================================
  def get_item_detail(self,item_code):
    it=sql("select item_name,brand,item_group,description,stock_uom from `tabItem` where name='%s'"%item_code)
    if it:
      ret = {
      'item_name'  : it and it[0][0] or '',
      'brand'      : it and it[0][1] or '',
      'item_group' : it and it[0][2] or '',
      'description': it and it[0][3] or '',
      'uom' : it and it[0][4] or ''
      }
      return cstr(ret)

  def validate(self):
    import string
    # Get Address
    # ======================================================================
    if (self.doc.address_line1)  or  (self.doc.address_line2) or (self.doc.city) or (self.doc.state) or (self.doc.country) or (self.doc.pincode):
      address =["address_line1", "address_line2", "city", "state", "country", "pincode"]
      comp_address=''
      for d in address:
        if self.doc.fields[d]:
          comp_address += self.doc.fields[d] + NEWLINE
      if self.doc.website:
        comp_address += "Website : "+ self.doc.website
      self.doc.address = comp_address
      
    if self.doc.status == 'Lead Lost' and not self.doc.order_lost_reason:
      msgprint("Please Enter Order Lost Reason")
      raise Exception
      
    if self.doc.source == 'Campaign' and not self.doc.campaign_name:
      msgprint("Please specify campaign name")
      raise Exception
  
    self.set_last_contact_date()
    
  
  def on_update(self):
    # Add to calendar
    # ========================================================================
    if self.doc.contact_date and self.doc.last_contact_date != self.doc.contact_date:
      if self.doc.contact_by:
        self.add_calendar_event()
        
#--------------Validation For Last Contact Date-----------------
  def set_last_contact_date(self):
    if not self.doc.contact_date_ref:
      self.doc.contact_date_ref=self.doc.contact_date
      self.doc.last_contact_date=self.doc.contact_date_ref
    if self.doc.contact_date_ref != self.doc.contact_date:
      if getdate(self.doc.contact_date_ref) < getdate(self.doc.contact_date):
        self.doc.last_contact_date=self.doc.contact_date_ref
      else:
        msgprint("Contact Date Cannot be before Last Contact Date")
        raise Exception
      set(self.doc, 'contact_date_ref',self.doc.contact_date)

  # Add to Calendar
  # ===========================================================================
  def add_calendar_event(self):
    ev = Document('Event')
    ev.description = 'Contact ' + cstr(self.doc.lead_name) + '.To Discuss : ' + cstr(self.doc.to_discuss)
    ev.event_date = self.doc.contact_date
    ev.event_hour = '10:00'
    ev.event_type = 'Public'
    ev.ref_type = 'Lead'
    ev.ref_name = self.doc.name
    ev.save(1)
      
    user_list = ['CRM Manager', 'CRM User']
    for d in user_list:
      ch = addchild(ev, 'event_individuals', 'Event User', 0)
      ch.person = d
      ch.save()


#-----------------Email-------------------------------------------- 
  def send_emails(self, email=[], subject='', message=''):
    if email:
      sender_email= sql("Select email from `tabProfile` where name='%s'"%session['user'])
      #msgprint(sender_email)
      if sender_email and sender_email[0][0]:
        attach_list=[]
        for at in getlist(self.doclist,'enquiry_attachment_detail'):
          if at.select_file:
            attach_list.append(at.select_file)
        cc_list=[]
        if self.doc.cc_to:
          for cl in (self.doc.cc_to.split(',')):
            cc_list.append(cl)
        sendmail(email, sender = sender_email[0][0], subject = subject , parts = [['text/html', message]],attach=attach_list)
        sendmail(cc_list, sender = sender_email[0][0], subject = subject , parts = [['text/html', message]],attach=attach_list)        
        msgprint("Mail Sent")
        self.add_in_follow_up(message,'Email')
      else:
        msgprint("Please enter your mail id in Profile")
        raise Exception
    else :
      msgprint("Recipient name not specified")
      raise Exception

  def sent_mail(self):
    if not self.doc.subject or not self.doc.message:
      msgprint("Please enter subject & message in their respective fields.")
    else:
      self.send_emails(self.doc.email_id, subject = self.doc.subject ,message = self.doc.message)

#---------------------- Add details in follow up table----------------
  def add_in_follow_up(self,message,type):
    import datetime
    child = addchild( self.doc, 'follow_up', 'Follow up', 1, self.doclist)
    child.date = datetime.datetime.now().date().strftime('%Y-%m-%d')
    child.notes = message
    child.follow_up_type = type
    child.save()

#-------------------SMS----------------------------------------------
  def send_sms(self):
    if not self.doc.sms_message or not self.doc.mobile_no:
      msgprint("Please enter mobile number in Overview Section and message in SMS Section ")
      raise Exception
    else:
      receiver_list = []
      if self.doc.mobile_no:
        receiver_list.append(self.doc.mobile_no)
      for d in getlist(self.doclist,'lead_sms_detail'):
        if d.other_mobile_no:
          receiver_list.append(d.other_mobile_no)
    
    if receiver_list:
      msgprint(get_obj('SMS Control', 'SMS Control').send_sms(receiver_list, self.doc.sms_message))
      self.add_in_follow_up(self.doc.sms_message,'SMS')

File: modules/CRM/Lead/Lead.js
cur_frm.cscript.onload = function(doc, cdt, cdn) {
  if (!doc.date){ 
  doc.date = date.obj_to_str(new Date());
  doc.contact_date=""
  }
}

cur_frm.cscript.refresh = function(doc, cdt, cdn) {

//On Saving Lead
//==============================================
    if(doc.__islocal != 1){
      unhide_field(['Create Contact']);
      unhide_field(['Create Enquiry']);
      unhide_field(['Create Customer']); 
      if(doc.status == 'Lead Lost') {
        unhide_field('order_lost_reason');
        hide_field(['Create Customer']);
        hide_field(['Create Contact']);
        hide_field(['Create Enquiry']);
      }
      else hide_field('order_lost_reason');
     }
  else{
    hide_field(['Create Customer']);
    hide_field(['Create Contact']);
    hide_field(['Create Enquiry']);
  }
}


// Client Side Triggers
// ===========================================================
// ************ Status ******************
cur_frm.cscript.status = function(doc, cdt, cdn){
  cur_frm.cscript.refresh(doc, cdt, cdn);
}

// *********** Country ******************
// This will show states belonging to country
cur_frm.cscript.country = function(doc, cdt, cdn) {
  var mydoc=doc;
  $c('runserverobj', args={'method':'check_state', 'docs':compress_doclist([doc])},
    function(r,rt){
      if(r.message) {
        var doc = locals[mydoc.doctype][mydoc.name];
        get_field(doc.doctype, 'state' , doc.name).options = r.message;
        refresh_field('state');
      }
    }
  );
}


// Convert Lead to Contact
// ===============================================================
cur_frm.cscript['Make Contact'] = function(doc){
  var cont = LocalDB.create('Contact');
  cont = locals['Contact'][cont];
  if (doc.type=='Client') {
    cont.contact_type='Company';
    cont.is_customer=1;
  }
  else if(doc.type=='Channel Partner' || doc.type=='Consultant'){
    cont.contact_type = 'Individual';
    cont.is_sales_partner=1;
  }
  cont.is_primary_contact='No';
  cont.first_name = doc.lead_name;
  cont.contact_no = doc.contact_no;
  cont.mobile_no = doc.mobile_no;
  cont.email_id = doc.email_id;
  loaddoc("Contact",cont.name);
}

// Create New File
// ===============================================================
cur_frm.cscript['Create New File'] = function(doc){
  new_doc("File");
}

//Trigger in Item Table
//===================================
cur_frm.cscript.item_code=function(doc,cdt,cdn){
  var d = locals[cdt][cdn];
  if (d.item_code) { get_server_fields('get_item_detail',d.item_code,'lead_item_detail',doc,cdt,cdn,1);}
}

// Create New Customer
// ===============================================================
cur_frm.cscript['Create Customer'] = function(doc){
  n = createLocal("Customer");
  $c('dt_map', args={
	  'docs':compress_doclist([locals["Customer"][n]]),
	  'from_doctype':'Lead',
	  'to_doctype':'Customer',
	  'from_docname':doc.name,
    'from_to_list':"[['Lead', 'Customer']]"
  }
  , function(r,rt) {
    loaddoc("Customer", n);
    }
    );
  }
  
// Create New Enquiry
// ===============================================================
cur_frm.cscript['Create Enquiry'] = function(doc){
  n = createLocal("Enquiry");
  $c('dt_map', args={
	  'docs':compress_doclist([locals["Enquiry"][n]]),
	  'from_doctype':'Lead',
	  'to_doctype':'Enquiry',
	  'from_docname':doc.name,
    'from_to_list':"[['Lead', 'Enquiry']]"
  }
  , function(r,rt) {
    loaddoc("Enquiry", n);
    }
    );
  }
  
// Create New Contact
// ===============================================================
cur_frm.cscript['Create Contact'] = function(doc){
  new_doc("Contact");
  }
  
// Fetching Campaign Name
//================================================================
cur_frm.cscript.source = function(doc, cdt, cdn){
 if (doc.source == 'Campaign'){
  unhide_field('campaign_name');
  }
 else {
  hide_field('campaign_name');
}
}